From 59067d5470c4ac304578c29006cb508934e347d6 Mon Sep 17 00:00:00 2001
From: pieterg <pieterg@gmx.com>
Date: Sun, 9 Oct 2011 12:02:18 +0200
Subject: [PATCH] skin: add support for relative coordinates

c = center of parent widget
e = dimension of parent widget (width / height)

(0,0)----------------------------(e,0)
|                                    |
|                                    |
|                                    |
|                                    |
(0,c)           (c,c)            (e,c)
|                                    |
|                                    |
|                                    |
|                                    |
(e,0)----------------------------(e,e)

e.g.:
position="c-50,c-50" size="100,100"
position="10,10" size="e-10,e-10"
---
 skin.py |   85 +++++++++++++++++++++++++++++++++++++++++++++-----------------
 1 files changed, 61 insertions(+), 24 deletions(-)

diff --git a/skin.py b/skin.py
index 19da203..12ca52d 100755
--- a/skin.py
+++ b/skin.py
@@ -70,31 +70,68 @@ profile("LoadSkinDefault")
 loadSkin('skin_default.xml')
 profile("LoadSkinDefaultDone")
 
-def evalPos(pos, wsize, ssize, scale):
-	if pos == "center":
-		pos = (ssize - wsize) / 2
+def parseCoordinate(str, e, size = 0):
+	str = str.strip()
+	if str == "center":
+		val = (e - size)/2
 	else:
-		pos = int(pos) * scale[0] / scale[1]
-	return int(pos)
+		sl = len(str)
+		l = 1
 
-def parsePosition(str, scale, desktop = None, size = None):
-	x, y = str.split(',')
-	
-	wsize = 1, 1
-	ssize = 1, 1
-	if desktop is not None:
-		ssize = desktop.size().width(), desktop.size().height()
-	if size is not None:
-		wsize = size.width(), size.height()
-
-	x = evalPos(x, wsize[0], ssize[0], scale[0])
-	y = evalPos(y, wsize[1], ssize[1], scale[1])
-
-	return ePoint(x, y)
+		if str[0] is 'e':
+			val = e
+		elif str[0] is 'c':
+			val = e/2
+		else:
+			val = 0;
+			l = 0
 
-def parseSize(str, scale):
+		if sl - l > 0:
+			if str[sl-1] is '%':
+				val += e * int(str[l:sl-1]) / 100
+			else:
+				val += int(str[l:sl])
+	if val < 0:
+		val = 0
+	return val
+
+def getParentSize(object, desktop):
+	size = eSize()
+	if object:
+		parent = object.getParent()
+		# For some widgets (e.g. ScrollLabel) the skin attributes are applied to
+		# a child widget, instead of to the widget itself. In that case, the parent
+		# we have here is not the real parent, but it is the main widget.
+		# We have to go one level higher to get the actual parent.
+		# We can detect this because the 'parent' will not have a size yet
+		# (the main widget's size will be calculated internally, as soon as the child
+		# widget has parsed the skin attributes)
+		if parent and parent.size().isEmpty():
+			parent = parent.getParent()
+		if parent:
+			size = parent.size()
+		elif desktop:
+			#widget has no parent, use desktop size instead for relative coordinates
+			size = desktop.size()
+	return size
+
+def parsePosition(str, scale, object = None, desktop = None, size = None):
+	x, y = str.split(',')
+	parentsize = eSize()
+	if object and (x[0] in ['c', 'e'] or y[0] in ['c', 'e']):
+		parentsize = getParentSize(object, desktop)
+	xval = parseCoordinate(x, parentsize.width(), size and size.width())
+	yval = parseCoordinate(y, parentsize.height(), size and size.height())
+	return ePoint(xval * scale[0][0] / scale[0][1], yval * scale[1][0] / scale[1][1])
+
+def parseSize(str, scale, object = None, desktop = None):
 	x, y = str.split(',')
-	return eSize(int(x) * scale[0][0] / scale[0][1], int(y) * scale[1][0] / scale[1][1])
+	parentsize = eSize()
+	if object and (x[0] in ['c', 'e'] or y[0] in ['c', 'e']):
+		parentsize = getParentSize(object, desktop)
+	xval = parseCoordinate(x, parentsize.width())
+	yval = parseCoordinate(y, parentsize.height())
+	return eSize(xval * scale[0][0] / scale[0][1], yval * scale[1][0] / scale[1][1])
 
 def parseFont(str, scale):
 	name, size = str.split(';')
@@ -137,9 +174,9 @@ def applySingleAttribute(guiObject, desktop, attrib, value, scale = ((1,1),(1,1)
 	# and set attributes
 	try:
 		if attrib == 'position':
-			guiObject.move(parsePosition(value, scale, desktop, guiObject.csize()))
+			guiObject.move(parsePosition(value, scale, guiObject, desktop, guiObject.csize()))
 		elif attrib == 'size':
-			guiObject.resize(parseSize(value, scale))
+			guiObject.resize(parseSize(value, scale, guiObject, desktop))
 		elif attrib == 'title':
 			guiObject.setTitle(_(value))
 		elif attrib == 'text':
-- 
1.7.4.1

